servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/GraphQLRequestHandler.kt

```diff
@@ -16,37 +16,139 @@
 import com.expediagroup.graphql.server.types.GraphQLResponse
+import com.expediagroup.graphql.server.types.GraphQLServerRequest
+import com.expediagroup.graphql.server.types.GraphQLServerResponse
 import graphql.GraphQL
+import graphql.execution.instrumentation.ChainedInstrumentation
+import graphql.execution.instrumentation.Instrumentation
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
 import kotlinx.coroutines.future.await
+import kotlinx.coroutines.supervisorScope
 
 open class GraphQLRequestHandler(
     private val graphQL: GraphQL,
-    private val dataLoaderRegistryFactory: DataLoaderRegistryFactory? = null
+    private val dataLoaderRegistryFactory: KotlinDataLoaderRegistryFactory? = null
 ) {
 
+    private val dataLoaderInstrumentationType: Class<Instrumentation>? =
+        graphQL.instrumentation?.let { instrumentation ->
+            when {
+                instrumentation is ChainedInstrumentation -> {
+                    instrumentation.instrumentations
+                        .firstOrNull(Instrumentation::isBatchDataLoaderInstrumentation)
+                        ?.javaClass
+                }
+                instrumentation.isBatchDataLoaderInstrumentation() -> instrumentation.javaClass
+                else -> null
+            }
+        }
+
     /**
      * Execute a GraphQL request in a non-blocking fashion.
      * This should only be used for queries and mutations.
      * Subscriptions require more specific server logic and will need to be handled separately.
      */
-    open suspend fun executeRequest(request: GraphQLRequest, context: GraphQLContext? = null, graphQLContext: Map<*, Any> = emptyMap<Any, Any>()): GraphQLResponse<*> {
-        // We should generate a new registry for every request
-        val dataLoaderRegistry = dataLoaderRegistryFactory?.generate()
-        val executionInput = request.toExecutionInput(context, dataLoaderRegistry, graphQLContext)
+    open suspend fun executeRequest(
+        graphQLRequest: GraphQLServerRequest,
+        context: GraphQLContext? = null,
+        graphQLContext: Map<*, Any> = emptyMap<Any, Any>()
+    ): GraphQLServerResponse =
+        when (graphQLRequest) {
+            is GraphQLRequest -> {
+                execute(graphQLRequest, context, graphQLContext, dataLoaderRegistryFactory?.generate())
+            }
+            is GraphQLBatchRequest -> {
+                if (graphQLRequest.containsMutation()) {
+                    executeBatchSequentially(graphQLRequest, context, graphQLContext)
+                } else {
+                    executeBatchConcurrently(graphQLRequest, context, graphQLContext)
+                }
+            }
+        }
 
+    private suspend fun execute(
+        graphQLRequest: GraphQLRequest,
+        context: GraphQLContext? = null,
+        graphQLContext: Map<*, Any> = emptyMap<Any, Any>(),
+        dataLoaderRegistry: KotlinDataLoaderRegistry? = null
```

Comment:
instead of `null` if it defaults to `dataLoaderRegistryFactory?.generate()` you would only need to pass it from batch async method

Reply:
i agree, and i think there is some value on passing the DataLoaderRegistry instance even if we need to execute sequentially, we are not going to batch but we can still use the `ValueCache` of the DataLoader, this to avoid retrieving data if it was already retrieved